openoa/utils/downloader.py

```diff
@@ -0,0 +1,428 @@
+        
+        
+        # download and unzip if the file doesn't exist
+        else:
+            
+            logger.info(f"\nDownloading: {file_name}")
+            logger.info(f"File size: {f['size']/(1024*1024):,.2f} MB")
+
+            download_file(url_file,outfile)
+
+            logger.info(f"Saved to: {outfile}\n")
+
+            if outfile.suffix == ".zip":
+                with ZipFile(outfile) as zipfile:
+                    zipfile.extractall(outfile_path)
+
+
+
+def get_era5(
+        lat: float,
+        lon: float,
+        save_pathname: str | Path,
+        save_filename: str,
+        ) -> pd.DataFrame:
+    """
+    Get ERA5 data directly from the CDS service, which requires registration on the CDS service.
+    See: https://cds.climate.copernicus.eu/api-how-to
+
+    Monthly 10m height data is demonstrated here, as hourly data takes too long to download, 
+    but could be amended and other CDS datasets also used (e.g. CERRA for Europe)
+
+    Args:
+        lat(:obj:`float`): Latitude in WGS 84 spatial reference system (decimal degrees)
+        lon(:obj:`float`): Longitude in WGS 84 spatial reference system (decimal degrees)
+        save_pathname(:obj:`str` | :obj:`Path`): The path where the downloaded reanalysis data will
+            be saved
+        save_filename(:obj:`str`): The file name used to save the downloaded reanalysis data
+
+    Returns:
+        df(:obj:`dataframe`): A dataframe containing time series of the requested reanalysis variables
+        Saved NetCDF annual ERA5 files
+        Saved ERA5 csv file
+    """
+
+    logger.info("Please note access to ERA5 data requires registration")
+    logger.info("Please see: https://cds.climate.copernicus.eu/api-how-to")
+
+    # set up cds-api client
+    try:
+        c = cdsapi.Client()
+    except Exception as e:
+        logger.error('Failed to make connection to cds: '+ str(e))
+        logger.error('Please see: https://cds.climate.copernicus.eu/api-how-to')
+        raise NameError(e)
+
+    # create save_pathname if it does not exist
+    save_pathname = Path(save_pathname).resolve()
+    if not save_pathname.exists():
+        save_pathname.mkdir()
+
+    # downloads all years from today back to the year 2000
+    now = datetime.datetime.now()
+    years = list(range(2000,now.year+1,1))
+
+    # get the data for the closest 9 nodes to the coordinates
+    node_spacing = 0.250500001*1
+
+    # download the data
+    for year in years:
+
+        outfile = save_pathname / f"{save_filename}_{year}.nc"
+
+        if year == now.year:
+            months = list(range(1,now.month,1))
+        else:
+            months = list(range(1,12+1,1))
+
+        # See: https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels-monthly-means?tab=form
+        # for formulating other requests from cds
+        if not outfile.is_file() or year==now.year:
+
+            logger.info(f"Downloading ERA5: {outfile}")
```

Comment:
After playing around in the notebook a bit, I think it would be best that we check for hte file first, then download if needed instead of automatically redownloading every time. The 2023 file downloads every time for me.

Reply:
Okay, I fixed this by limiting the dates monthly rather than yearly, so that if new data becomes available during the year then this does still get downloaded. 

I also updated the function so that it only returns the specified date range, as previously it would provide all downloaded data.

Reply:
Thank  you!